File: examples/iotApp/Light.cpp
//******************************************************************
//
// Copyright 2016 Intel Corporation All Rights Reserved.
//
//-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
//-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

#include "Light.h"
#include <OCPlatform.h>
#include "IotivityUtility.h"

using namespace std;
using namespace OC;

Light::Light() :
    m_resource(),
    m_supported_services_known(false),
    m_lightSupportsBrightness(false),
    m_lightSupportsSwitch(false),
    m_powerState(false),
    m_eCode(0)
{
}

Light::Light(OCResource::Ptr resource) :
    m_resource(resource),
    m_supported_services_known(false),
    m_lightSupportsBrightness(false),
    m_lightSupportsSwitch(false),
    m_powerState(false),
    m_eCode(0)
{
    init();
}

Light::~Light()
{
    unique_lock<mutex> lock(m_mutex);
}

Light::Light( const Light &other )
{
    *this = other;
}

bool Light::init()
{
    unique_lock<mutex> lock(m_mutex);
    while (!m_supported_services_known)
    {
        onGetServicesCB = bind(&Light::onGetServices, this, placeholders::_1, placeholders::_2,
                               placeholders::_3);
        m_resource->get(QueryParamsMap(), onGetServicesCB);
        if (m_cv.wait_for(lock, chrono::seconds(MAX_WAIT_TIME_FOR_BLOCKING_CALL)) == cv_status::timeout)
        {
            cerr << "Remote device failed to respond to the request." << endl;
        }
    }
    return true;
}
Light &Light::operator=(const Light &other)
{
    unique_lock<mutex> lock(m_mutex);
    // check for self-assignment
    if (this != &other)
    {
        onGetServicesCB = bind(&Light::onGetServices, this, placeholders::_1, placeholders::_2,
                               placeholders::_3);
        m_resource = other.m_resource;
        m_supported_services_known = other.m_supported_services_known;
        m_lightSupportsBrightness = other.m_lightSupportsBrightness;
        m_lightSupportsSwitch = other.m_lightSupportsSwitch;
        m_powerState = other.m_powerState;
        m_binarySwitch = other.m_binarySwitch;
        m_brightness = other.m_brightness;
        // do not copy the mutex or condition_variable
    }
    return *this;
}

void Light::toggle()
{
    if (isOn())
    {
        turnOn(false);
    }
    else
    {
        turnOn(true);
    }
}

bool Light::turnOn(bool isOn)
{
    return m_binarySwitch.turnOn(isOn);
}

bool Light::isOn()
{
    return m_binarySwitch.isOn();
}

void Light::isOnAsync(GetCallback isOnCB)
{
    m_binarySwitch.isOnAsync(isOnCB);
}

void Light::turnOnAsync(bool isOn, PostCallback turnOnCB) const
{
    m_binarySwitch.turnOnAsync(isOn, turnOnCB);
}

int Light::getBrightness()
{
    return m_brightness.getBrightness();
}

bool Light::setBrightness(int brightness)
{
    return m_brightness.setBrightness(brightness);
}

void Light::getBrightnessAsync(GetCallback getBrightnessCB)
{
    m_brightness.getBrightnessAsync(getBrightnessCB);
}

void Light::setBrightnessAsync(int brightness, PostCallback setBrightnessCB)
{
    m_brightness.setBrightnessAsync(brightness, setBrightnessCB);
}

bool Light::operator<(const Light &other) const
{
    return ((*m_resource) < (*(other.m_resource)));
}

void Light::onGetServices(const HeaderOptions &headerOptions, const OCRepresentation &rep,
                          const int eCode)
{
    AttributeValue links;
    if (rep.hasAttribute("links"))
    {
        if (rep.getAttributeValue("links", links))
        {
            try
            {
                auto linkRepresentation = boost::get<vector<OCRepresentation> >(links);
                for (auto l : linkRepresentation)
                {
                    if (l.hasAttribute("rel"))
                    {
                        string rel;
                        if (l.getValue("rel", rel))
                        {
                            cout << "rel: " << rel << endl;
                            if ("contains" == rel)
                            {
                                if (l.hasAttribute("rt"))
                                {
                                    string rt;
                                    if (l.getValue("rt", rt))
                                    {
                                        cout << "rt: " << rt << endl;
                                        if ("oic.r.switch.binary" == rt)
                                        {
                                            if (l.hasAttribute("href"))
                                            {
                                                string href;
                                                l.getValue("href", href);
                                                cout << "href: " << href << endl;
                                                m_lightSupportsSwitch = true;
                                                OCResource::Ptr resource = OCPlatform::constructResourceObject(m_resource->host(), href,
                                                                           m_resource->connectivityType(), true, {rt}, {"oic.if.baseline", "oic.if.a"});
                                                m_binarySwitch = BinarySwitch(resource);
                                            }
                                        }
                                        else if ("oic.r.light.brightness" == rt)
                                        {
                                            if (l.hasAttribute("href"))
                                            {
                                                string href;
                                                l.getValue("href", href);
                                                cout << "href: " << href << endl;
                                                m_lightSupportsBrightness = true;
                                                OCResource::Ptr resource = OCPlatform::constructResourceObject(m_resource->host(), href,
                                                                           m_resource->connectivityType(), true, {rt}, {"oic.if.baseline", "oic.if.a"});
                                                m_brightness = Brightness(resource);
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
            catch (boost::bad_get &e)
            {
                cout << e.what() << endl;
                cout << "falure to getValue for links" << endl;
                cout << rep.getValueToString("links");
            }
        }
        else
        {
            cout << "falure to getValue for links" << endl;
            cout << rep.getValueToString("links");
        }
    }
    m_supported_services_known = true;
    m_cv.notify_one();
}

File: android/examples/upnpclient/src/main/java/org/iotivity/base/examples/Light.java
/*
 * //******************************************************************
 * //
 * // Copyright 2016 Intel Corporation All Rights Reserved.
 * //
 * //-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
 * //
 * // Licensed under the Apache License, Version 2.0 (the "License");
 * // you may not use this file except in compliance with the License.
 * // You may obtain a copy of the License at
 * //
 * //      http://www.apache.org/licenses/LICENSE-2.0
 * //
 * // Unless required by applicable law or agreed to in writing, software
 * // distributed under the License is distributed on an "AS IS" BASIS,
 * // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * // See the License for the specific language governing permissions and
 * // limitations under the License.
 * //
 * //-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
 */

package org.iotivity.base.examples;

import org.iotivity.base.OcException;
import org.iotivity.base.OcRepresentation;

/**
 * Light
 * <p/>
 * This class is used by UpnpClientActivity to create an object representation of a remote light device
 * and update the values depending on the server response
 */
public class Light extends Device {

    public static final String OIC_TYPE_DEVICE_LIGHT = "oic.d.light";
    public static final String OCF_OIC_URI_PREFIX_LIGHT = "/ocf/light/";
    public static final String UPNP_OIC_URI_PREFIX_LIGHT = "/upnp/light/";
    public static final String OIC_URI_PREFIX_LIGHT = "/light/";


    private BinarySwitch mBinarySwitch;
    private Brightness mBrightness;
    private Configuration mConfig;

    private String mName = "Unknown";
    private boolean mState = BinarySwitch.DEFAULT_VALUE;
    private int mLightLevel = Brightness.DEFAULT_BRIGHTNESS;

    public Light() {
        super();
    }

    public void setOcRepresentation(OcRepresentation rep) throws OcException {
        super.setOcRepresentation(rep);
        if (rep.hasAttribute(Configuration.NAME_KEY)) {
            Object obj = rep.getValue(Configuration.NAME_KEY);
            if (obj instanceof String) {
                mName = (String) obj;
            }
        }
        if (rep.hasAttribute(BinarySwitch.VALUE_KEY)) {
            Object obj = rep.getValue(BinarySwitch.VALUE_KEY);
            if (obj instanceof Boolean) {
                mState = (Boolean) obj;
            }
        }
        if (rep.hasAttribute(Brightness.BRIGHTNESS_KEY)) {
            Object obj = rep.getValue(Brightness.BRIGHTNESS_KEY);
            if (obj instanceof Integer) {
                mLightLevel = (Integer) obj;
            }
        }
    }

    public OcRepresentation getOcRepresentation() throws OcException {
        OcRepresentation rep = super.getOcRepresentation();
        rep.setValue(Configuration.NAME_KEY, mName);
        rep.setValue(BinarySwitch.VALUE_KEY, mState);
        rep.setValue(Brightness.BRIGHTNESS_KEY, mLightLevel);
        return rep;
    }

    public BinarySwitch getBinarySwitch() {
        return mBinarySwitch;
    }

    public void setBinarySwitch(BinarySwitch binarySwitch) {
        mBinarySwitch = binarySwitch;
    }

    public Brightness getBrightness() {
        return mBrightness;
    }

    public void setBrightness(Brightness brightness) {
        mBrightness = brightness;
    }

    public Configuration getConfiguration() {
        return mConfig;
    }

    public void setConfiguration(Configuration config) {
        mConfig = config;
    }

    public String getName() {
        if (mConfig != null) {
            mName = mConfig.getName();
        }

        return mName;
    }

    public void setName(String name) {
        if (mConfig != null) {
            mConfig.setName(name);

        } else {
            mName = name;
        }
    }

    public boolean getState() {
        if (mBinarySwitch != null) {
            mState = mBinarySwitch.getValue();
        }

        return mState;
    }

    public void setState(boolean state) {
        if (mBinarySwitch != null) {
            mBinarySwitch.setValue(state);

        } else {
            mState = state;
        }
    }

    public int getLightLevel() {
        if (mBrightness != null) {
            mLightLevel = mBrightness.getBrightness();
        }

        return mLightLevel;
    }

    public void setLightLevel(int lightLevel) {
        if (mBrightness != null) {
            mBrightness.setBrightness(lightLevel);

        } else {
            mLightLevel = lightLevel;
        }
    }

    @Override
    public String toString() {
        String binarySwitchAsString = mBinarySwitch != null ? mBinarySwitch.toString() : "binary switch is null";
        String brightnessAsString = mBrightness != null ? mBrightness.toString() : "brightness is null";
        String configAsString = mConfig != null ? mConfig.toString() : "configuration is null";

        if (mConfig != null) {
            return "[" + super.toString() +
                    ", " + binarySwitchAsString +
                    ", " + brightnessAsString +
                    ", " + configAsString + "]";
        } else {
            return "[" + super.toString() +
                    ", " + Configuration.NAME_KEY + ": " + mName +
                    ", " + BinarySwitch.VALUE_KEY + ": " + mState +
                    ", " + Brightness.BRIGHTNESS_KEY + ": " + mLightLevel + "]";
        }
    }
}
